File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/express/add_new_card/OfflineMethods.java
package com.mercadopago.android.px.internal.features.express.add_new_card;

import android.support.annotation.NonNull;
import com.mercadopago.android.px.internal.base.MvpView;
import com.mercadopago.android.px.internal.features.pay_button.PayButton;
import com.mercadopago.android.px.internal.viewmodel.AmountLocalized;

public interface OfflineMethods {

    interface OffMethodsView extends MvpView, PayButton.Handler {

        void updateTotalView(@NonNull final AmountLocalized amountLocalized);

        void onSlideSheet(final float offset);

        void startKnowYourCustomerFlow(@NonNull final String flowLink);
    }

    interface Actions {

        void updateModel();

        void selectMethod(@NonNull final OfflineMethodItem selectedItem);

        void onBack();
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/express/offline_methods/OfflineMethods.kt
package com.mercadopago.android.px.internal.features.express.offline_methods

import androidx.lifecycle.LiveData
import com.mercadopago.android.px.internal.features.pay_button.PayButton
import com.mercadopago.android.px.internal.viewmodel.AmountLocalized
import com.mercadopago.android.px.model.ExpressMetadata
import com.mercadopago.android.px.model.OfflinePaymentType
import com.mercadopago.android.px.model.internal.PaymentConfiguration
import com.mercadopago.android.px.model.internal.Text

internal interface OfflineMethods {

    interface View : PayButton.Handler {
        fun showExpanded()
        fun showCollapsed()
    }

    interface ViewModel {
        val deepLinkLiveData: LiveData<String>
        fun onSheetShowed()
        fun onViewLoaded(): LiveData<Model>
        fun onMethodSelected(selectedItem: OfflineMethodItem)
        fun onPrePayment(callback: PayButton.OnReadyForPaymentCallback)
        fun onBack()
        fun onPaymentExecuted(configuration: PaymentConfiguration)
    }

    interface OnMethodSelectedListener {
        fun onItemSelected(selectedItem: OfflineMethodItem)
    }

    data class Model(
        val bottomDescription: Text?,
        val amountLocalized: AmountLocalized,
        val offlinePaymentTypes: List<OfflinePaymentType>)

    companion object {
        @JvmStatic
        fun shouldLaunch(expressMetadataList: List<ExpressMetadata>): Boolean {
            return expressMetadataList.filter { express -> express.status.run { isEnabled } }
                .run { size == 1 && with(get(0)) { isOfflineMethods && !isNewCard } }
        }
    }
}